imagebuildah/build.go

```diff
@@ -1251,3 +1267,56 @@ func (b *Executor) deleteSuccessfulIntermediateCtrs() error {
 	}
 	return lastErr
 }
+
+// preprocessDockerfileContents runs CPP(1) in preprocess-only mode on the input
+// dockerfile content and will use ctxDir as the base include path.
+//
+// Note: we cannot use cmd.StdoutPipe() as cmd.Wait() closes it.
+func preprocessDockerfileContents(r io.ReadCloser, ctxDir string) (rdrCloser *io.ReadCloser, err error) {
+	cppPath := "/usr/bin/cpp"
+	if _, err = os.Stat(cppPath); err != nil {
+		if os.IsNotExist(err) {
+			err = errors.Errorf("error: Dockerfile.in support requires %s to be installed", cppPath)
+		}
+		return
+	}
+
+	stdout := bytes.Buffer{}
+	stderr := bytes.Buffer{}
+
+	cmd := exec.Command(cppPath, "-E", "-iquote", ctxDir, "-")
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	pipe, err := cmd.StdinPipe()
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			pipe.Close()
+		}
+	}()
+
+	if err = cmd.Start(); err != nil {
+		return
```

Comment:
I think  you need to return err here now.

THen the err will be set. and you don't have to worry about err.  All returns should return err.

Reply:
This won't compile as we have two return values. Or am I missing something? Can't we just do a `defer pipe.Close()` and one explicit one before the `cmd.Wait()`? In the worst case, we have one redundant Close() but the code will be easier to understand (imho). Personally, I find that named return values complicate the maintainability of code.

Reply:
Sorry 
return nil, err

The values of the returns get populated when you return before the defer gets called.

So you can do 
return nil, errors.Error("foobar") 
and when the defer gets called the err == foobar.